Add Helm instructions and Daemonset cleanup (#11063)
DataDog/documentation: content/en/agent/cluster_agent/external_metrics.md

    # Enable the metricsProvider to be able to scale based on metrics in Datadog
    metricsProvider:
      # clusterAgent.metricsProvider.enabled
      # Set this to true to enable Metrics Provider
      enabled: true
  ```

This automatically updates the necessary RBAC configurations as well as sets up the corresponding `Service` and `APIService` for Kubernetes to utilize.

[1]: https://github.com/DataDog/helm-charts/blob/master/charts/datadog/values.yaml
{{% /tab %}}
{{% tab "Daemonset" %}}

### Custom metrics server

To enable the Custom Metrics Server, first follow the instructions to [set up the Datadog Cluster Agent][1] within your cluster. Once you have verified a succesful base deployment, edit your `Deployment` manifest for the Datadog Cluster Agent with the following steps:

1. Set `DD_EXTERNAL_METRICS_PROVIDER_ENABLED` environment variable to `true`
2. Ensure you have **both** your environment variables `DD_APP_KEY` and `DD_API_KEY` set
3. Ensure you have your `DD_SITE` environment variable set to your Datadog site: {{< region-param key="dd_site" code="true" >}}. It defaults to the `US` site `datadoghq.com`.

### Register the external metrics provider service

> Once the Datadog Cluster Agent is up and running you will need to apply some additional RBAC policies and setup the `Service` to route the corresponding requests.

Reviewer: ```suggestion
Once the Datadog Cluster Agent is up and running, apply some additional RBAC policies and set up the `Service` to route the corresponding requests.
```

1. Create a `Service` named `datadog-custom-metrics-server`, exposing the port `443` with the following `custom-metric-server.yaml` manifest:

    ```yaml
    kind: Service
    apiVersion: v1
    metadata:
      name: datadog-custom-metrics-server
    spec:
      selector:
        app: datadog-cluster-agent
      ports:
      - protocol: TCP
        port: 443
        targetPort: 443
    ```
    **Note:** The Cluster Agent by default is expecting these requests over port `443`. However, if your Cluster Agent `Deployment` has set the environment variable `DD_EXTERNAL_METRICS_PROVIDER_PORT` to some other port value, change the `targetPort` of this `Service` accordingly.

    Apply this `Service` by running `kubectl apply -f custom-metric-server.yaml`
2. Download the [`rbac-hpa.yaml` RBAC rules file][2].

> 3. Register the Cluster Agent as an External Metrics Provider by applying this file:

Reviewer: ```suggestion
3. Register the Cluster Agent as an external metrics provider by applying this file:
```

- The maximum number of replicas created is `3`, and the minimum is `1`.
- The metric used is `nginx.net.request_per_s`, and the scope is `kube_container_name: nginx`. This metric format corresponds to the Datadog one.

Every 30 seconds, Kubernetes queries the Datadog Cluster Agent to get the value of this metric and autoscales proportionally if necessary. For advanced use cases, it is possible to have several metrics in the same HPA. As you can see [in the Kubernetes horizontal pod autoscaling documentation][5], the largest of the proposed values is the one chosen.

**Note**: Running multiple Cluster Agents will raise API usage. The Datadog Cluster Agent completes 120 calls per hour for approximately 45 HPA objects in Kubernetes. Running more than 45 HPAs will increase the number of calls when fetching metrics from within the same org.

## Autoscaling with custom queries using DatadogMetric CRD

> You can autoscale on a Datadog query by using  the `DatadogMetric` [Custom Resource Definition (CRD)][6] and Datadog Cluster Agent versions `1.7.0` or above. This is a more flexible approach and will allow you to scale with the exact Datadog query you would use in app.

Reviewer: ```suggestion
You can autoscale on a Datadog query by using  the `DatadogMetric` [Custom Resource Definition (CRD)][6] and Datadog Cluster Agent versions `1.7.0` or above. This is a more flexible approach and allows you to scale with the exact Datadog query you would use in-app.
```

**Note**: While the query is arbitrary, the start and end times are still set at `Now() - 5 minutes` and `Now()`

### Set up Datadog Cluster Agent to use DatadogMetric
{{< tabs >}}
{{% tab "Helm" %}}

> To activate usage of the `DatdogMetric` CRD update your [datadog-values.yaml][1] Helm configuration to set `clusterAgent.metricsProvider.useDatadogMetrics` to `true`.  Then redeploy your Datadog Helm chart:

Reviewer: ```suggestion
To activate usage of the `DatdogMetric` CRD update your [datadog-values.yaml][1] Helm configuration to set `clusterAgent.metricsProvider.useDatadogMetrics` to `true`. Then redeploy your Datadog Helm chart:
```

  ```yaml
  clusterAgent:
    enabled: true
    metricsProvider:
      enabled: true
      # clusterAgent.metricsProvider.useDatadogMetrics
      # Enable usage of DatadogMetric CRD to autoscale on arbitrary Datadog queries
      useDatadogMetrics: true
  ```

> **Note:** This will automatically attempt to install the `DatadogMetric` CRD. If that CRD already exists prior to the initial Helm instalation it may conflict.

Reviewer: ```suggestion
**Note:** This will automatically attempts to install the `DatadogMetric` CRD. If that CRD already exists prior to the initial Helm installation, it may conflict.
```

> This automatically updates the necessary RBAC files and will direct the Cluster Agent to manage these HPA queries through these `DatadogMetric` resources.

Reviewer: ```suggestion
This automatically updates the necessary RBAC files and directs the Cluster Agent to manage these HPA queries through these `DatadogMetric` resources.
```

### Troubleshooting

> The Datadog Cluster Agent will take care of updating the `status` subresource of all `DatadogMetric` resources to reflect results from queries to Datadog. This is the main source of information to understand what happens if something is failing. You can run the following to get this information outputted for you:

Reviewer: ```suggestion
The Datadog Cluster Agent takes care of updating the `status` subresource of all `DatadogMetric` resources to reflect results from queries to Datadog. This is the main source of information to understand what happens if something is failing. You can run the following to get this information outputted for you:
```
